lib/helper.js

```diff
@@ -0,0 +1,218 @@
+    Object.assign(newJob.settings, oldJob.settings || {});
+    newJob.image = oldJob.image || newJob.image;
+
+    if (oldJob.requires) {
+        newJob.requires = [].concat(oldJob.requires);
+    } // otherwise, keep it the same, or don't set if it wasn't set
+
+    if (oldJob.blockedBy) {
+        newJob.blockedBy = [].concat(oldJob.blockedBy);
+    }
+
+    if (oldJob.freezeWindows) {
+        newJob.freezeWindows = [].concat(oldJob.freezeWindows);
+    }
+
+    if (oldJob.cache || oldJob.cache === false) {
+        newJob.cache = oldJob.cache;
+    }
+
+    // Merge secrets
+    const newSecrets = newJob.secrets || [];
+    const oldSecrets = oldJob.secrets || [];
+
+    // Merge sourcePaths
+    let newsourcePaths = newJob.sourcePaths || [];
+    let oldsourcePaths = oldJob.sourcePaths || [];
+
+    newsourcePaths = Array.isArray(newsourcePaths) ? newsourcePaths : [newsourcePaths];
+    oldsourcePaths = Array.isArray(oldsourcePaths) ? oldsourcePaths : [oldsourcePaths];
+
+    // remove duplicate
+    newJob.secrets = [...new Set([...newSecrets, ...oldSecrets])];
+    newJob.sourcePaths = [...new Set([...newsourcePaths, ...oldsourcePaths])];
+
+    // Use "order" to get steps, ignore all other steps; current template
+    // has precedence over external template
+    if (fromTemplate && oldJob.order) {
+        let stepName;
+        const mergedSteps = [];
+
+        // Convert steps from array to object for faster lookup
+        const oldSteps = convertFromArrayToObject(oldJob.steps);
+        const newSteps = convertFromArrayToObject(newJob.steps);
+
+        for (let i = 0; i < oldJob.order.length; i += 1) {
+            stepName = oldJob.order[i];
+
+            if (oldSteps && oldSteps[stepName]) {
+                mergedSteps.push({ [stepName]: oldSteps[stepName] });
+            } else if (newSteps && newSteps[stepName]) {
+                mergedSteps.push({ [stepName]: newSteps[stepName] });
+            } else {
+                warnings = warnings.concat(
+                    // eslint-disable-next-line max-len
+                    `${stepName} step in cannot be found in current template or ${oldJob.template} template; skipping`
+                );
+            }
+        }
+
+        newJob.steps = mergedSteps;
+    // Basic step merge with template
+    } else if (fromTemplate && oldJob.steps) {
+        let stepName;
+        let preStepName;
+        let postStepName;
+        const mergedSteps = [];
+        const teardownSteps = [];
+
+        // convert steps from oldJob from array to object for faster lookup
+        const oldSteps = oldJob.steps.reduce((obj, item) => {
+            const key = Object.keys(item)[0];
+
+            if (key.startsWith('teardown-')) {
+                teardownSteps.push(key);
+            }
+
+            obj[key] = item[key];
+
+            return obj;
+        }, {});
+
+        for (let i = 0; i < newJob.steps.length; i += 1) {
```

Comment:
I don't think child template should have pre/post, if necessary people should just use `order` field.

Reply:
I'm okay with your proposal
@jithine what do you think?

Reply:
order should be optional.
1.  If order is not used, then regular step merging should happen.
2. If order is used, only order should be used to determine step order, no need for pre/post
3. teardown- should continue to work as it is currently. at the end of all user steps.

Reply:
1. This is already in my PR
2. This is already in my PR as well, though I didn't explicitly prevent pre/post steps
3. Added support for teardown-